gsoc-2025/semantic-routines.md

```diff
@@ -0,0 +1,29 @@
+---
+parent: DLang GSoC 2025 Project Ideas
+nav_order: 1
+---
+
+# Separate Semantic Routines From AST Nodes
+
+**Mentor:** [Razvan Nitu](razvan.nitu1305@gmail.com)
+
+| Spec | Details |
+|-|-|
+| **Difficulty Level** | easy-medium |
+| **Project Duration** | 200 hours |
+| **Number of Contributors** | 1 |
+| **Prerequisites** | Familiarity with compiler organization, visitor pattern, object oriented programming |
+
+## Description
+
+In the DMD compiler codebase, AST nodes are defined as classes within various files. The ideal structure for these nodes is to have minimal fields and methods focused solely on field queries. However, the current state of the DMD frontend deviates from this ideal. AST nodes are laden with numerous methods that either perform or are dependent on semantic analysis. Furthermore, many AST node files contain free functions related to semantic analysis. Our objective is to decouple AST nodes from these functions, both directly and indirectly.
+
+This project helps advance the development of the compiler library by creating a clear separation between compilation phases.
+
+This project is ideal for someone that has no prior experience with compilers but wants to understand how a compiler works by doing valuable work.
```

Comment:
This statement contradicts the prerequisite in the table at the top: "Familiarity with compiler organization" and is meaningless. Highlight the value of the work by showing how it is useful for D (give D-Scanner as an example).